### ISSUE TITLE
Azure/go-autorest will soon be unsupported

### ISSUE THREAD

**Issue by mboersma** [kubernetes-sigs/cluster-api-provider-azure#2974]
/kind deprecation
/priority important-soon

Related to the SDK v2 transition, the `Azure/go-autorest` library will go [out of support](https://github.com/Azure/go-autorest#note-the-modules-in-this-repo-will-go-out-of-support-by-march-31-2023--additional-information-can-be-found-here) on March 31, 2023. CAPZ should find alternatives to those modules–see [here](https://github.com/kubernetes-sigs/cloud-provider-azure/pull/2983) for an example of replacing `to.StringPtr()` and friends.

**Comment by dtzar** [comment]
@mboersma - isn't this a duplicate of https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2670 ?

**Comment by mboersma** [comment]
@dtzar they're related but not the same: Azure/azure-sdk-for-go and Azure/go-autorest. CAPZ's direct usage of go-autorest can probably be addressed as a separate task, similarly to kubernetes-sigs/cloud-provider-azure#2983.

**Comment by CecileRobertMichon** [comment]
@mboersma are you interested in working on this or should we mark it as help wanted?

**Comment by mboersma** [comment]
/assign

I'm psyched to work on it but since it's also intertwined with our async framework it will be more work than just replacing our `to.StringPtr()`-type calls.

**Comment by CecileRobertMichon** [comment]
Should we maybe start with the `to.*Ptr` type calls since we'll have to do that regardless of SDK track 2 vs. ASO stuff? that can be its own PR. Then we can evaluate moving async services separately.

**Comment by dtzar** [comment]
@mboersma should this be attached to #3409 parent?

**Comment by mboersma** [comment]
> @mboersma should this be attached to https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3409 parent?

Yes, good call. Although really it will probably require all of ASO to be implemented as well before go-autorest is literally not used.

**Comment by mboersma** [comment]
/unassign

I've done the up-front work that was possible here. Let's revisit this once we've switched all over to SDKv2 + ASO and see what remains from go-autorest.

**Comment by mboersma** [comment]
/milestone next

I took a look around, and this will be a fair amount of work even after SDK v2 is fully integrated. go-autorest is woven into our opentelemetry tracing, we use autorest errors and objects as test data, and the go-autorest auth struct is everywhere.

**Comment by mboersma** [comment]
See also #4208 which removes a lot of the code in question.

**Comment by dtzar** [comment]
@mboersma can we complete the rest of this work for 1.13 and if so, what is remaining?

**Comment by mboersma** [comment]
@dtesar there are still a few vestigial uses of this library in tests and in some auth code, and in the latter case there isn't an obvious replacement for CAPZ's use of things like `auth.Environment`.

I don't think there's particular urgency to this, however: the uses are confined and the code in question isn't going to break. I'm hopeful to make some progress on this for v1.13, but there may be higher priorities.

**Comment by mboersma** [comment]
/unassign

I won't be able to address this, at last for a couple weeks. If anyone else wants to pick it up, that would be 💯.
